request.go

```diff
@@ -0,0 +1,251 @@
+package gomatrixserverlib
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/matrix-org/util"
+	"golang.org/x/crypto/ed25519"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// A MatrixRequest is a request to send to remote server or a request received
+// from a remote server.
+// Matrix requests are signed by building a JSON object and signing it
+type MatrixRequest struct {
```

Comment:
`FederationRequest` may be clearer given this has nothing to do with CS API and it could easily be conflated.